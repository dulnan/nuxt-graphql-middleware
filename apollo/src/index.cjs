const jiti = require('jiti')(__filename)

jiti('./app.ts')
