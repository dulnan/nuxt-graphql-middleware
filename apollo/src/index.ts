import { ApolloServer } from '@apollo/server'
import { expressMiddleware } from '@apollo/server/express4'
import { ApolloServerPluginDrainHttpServer } from '@apollo/server/plugin/drainHttpServer'
import express from 'express'
import http from 'http'
import cors from 'cors'
import bodyParser from 'body-parser'
import graphqlUploadExpress from 'graphql-upload/graphqlUploadExpress.mjs'
import GraphQLUpload, { FileUpload } from 'graphql-upload/GraphQLUpload.mjs'
import { GraphQLError } from 'graphql'
import data from './data.json' assert { type: 'json' }
import type { Readable } from 'stream'
import { v4 as uuidv4 } from 'uuid'

const BASIC_LOGGING: any = {
  requestDidStart(requestContext) {
    console.log('request started')
    console.log(requestContext.request.query)
    console.log(requestContext.request.variables)
    return {
      didEncounterErrors(requestContext) {
        console.log(
          'an error happened in response to query ' +
            requestContext.request.query,
        )
        console.log(requestContext.errors)
      },
    }
  },

  willSendResponse(requestContext) {
    console.log('response sent', requestContext.response)
  },
}

type UploadedFile = {
  id: string
  name: string
  content: string
}

type FormSubmissionDocument = {
  name: string
  file: UploadedFile
}

type FormSubmission = {
  id: string
  firstName: string
  lastName: string
  documents: FormSubmissionDocument[]
}

let users = []
let idIncrement = 0
let files: UploadedFile[] = []
let formSubmissions: FormSubmission[] = []

function initState() {
  users = [...data]
  idIncrement = users.length
  files = []
  formSubmissions = []
}

initState()

function getId() {
  idIncrement++
  return idIncrement
}

const typeDefs = `#graphql
  scalar Upload
  type User {
    id: Int!
    firstName: String!
    lastName: String!
    email: String!
    description: String
    dateOfBirth: String
    friends: [User]
  }

  input UserData {
    firstName: String!
    lastName: String!
    email: String!
    description: String
    dateOfBirth: String
  }

  type TestFetchOptions {
    headerClient: String
    headerServer: String
  }

type DataForLayer {
text: String
}

  type Query {
    users: [User!]!
    userById(id: ID!): User
    testFetchOptions: TestFetchOptions
    getRequestHeader(name: String!): String
    getError: Boolean
    getSubmissions: [FormSubmission]
    getCurrentTime: String
    dataForLayer: DataForLayer
  }

  type UploadedFile {
    id: String!
    name: String!
    content: String!
  }

  type FormSubmissionDocument {
    name: String
    file: UploadedFile!
  }

  type FormSubmission {
    id: String!
    firstName: String
    lastName: String
    documents: [FormSubmissionDocument]
  }

  type Mutation {
    createUser(user: UserData!): User!
    deleteUser(id: Int!): Boolean
    initState: Boolean!
    triggerError: Boolean
    uploadFile(file: Upload): Boolean!
    submitForm(input: FormSubmissionInput!): Boolean!
  }

  input FormSubmissionDocumentsInput {
    name: String
    file: Upload!
  }

  input FormSubmissionInput {
    firstName: String
    lastName: String
    documents: [FormSubmissionDocumentsInput]
  }
`

function streamToString(stream: Readable): Promise<string> {
  const chunks = []
  return new Promise((resolve, reject) => {
    stream.on('data', (chunk) => chunks.push(Buffer.from(chunk)))
    stream.on('error', (err) => reject(err))
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')))
  })
}

const resolvers = {
  Query: {
    users: () => {
      return users
    },
    getCurrentTime: () => {
      return new Date()
    },
    dataForLayer: () => {
      return { text: 'This is data for the layer page.' }
    },
    getSubmissions: () => {
      return formSubmissions
    },
    userById: (_: any, args: any) => {
      const id = parseInt(args.id)
      return users.find((v) => v.id === id)
    },
    testFetchOptions: (_parent: any, _args: any, context: any) => {
      return {
        headerClient: context.headerClient,
        headerServer: context.headerServer,
      }
    },
    getRequestHeader: (_parent: any, args: any, context: any) => {
      console.log('*'.repeat(50))
      console.log(args)
      console.log(context.headers)
      console.log('*'.repeat(50))
      return context.headers[args.name]
    },
    getError: () => {
      throw new GraphQLError('Something is wrong with your data.', {
        extensions: {
          code: 'WRONG_DATA',
        },
      })
    },
  },
  User: {
    friends: () => {
      return []
    },
  },
  Mutation: {
    createUser: (_: any, args: any) => {
      const user = { id: getId(), ...args.user }
      users.push(user)
      return user
    },
    deleteUser: (_: any, args: any) => {
      users = users.filter((v) => v.id !== args.id)
      return true
    },
    initState: () => {
      initState()
      return true
    },

    triggerError: () => {
      throw new GraphQLError('Something is wrong with your data.', {
        extensions: {
          code: 'WRONG_DATA',
        },
      })
    },

    uploadFile: async (_, { file }) => {
      const { filename, createReadStream } = await file
      console.log(`Uploading ${filename}...`)
      const stream = createReadStream()
      const content = await streamToString(stream)
      files.push({ id: uuidv4(), name: filename, content })
      return true
    },

    submitForm: async (_, { input }) => {
      const firstName = input.firstName
      const lastName = input.lastName
      const documents: FormSubmissionDocument[] = []

      for (let i = 0; i < input.documents.length; i++) {
        const doc = input.documents[i]
        const file: FileUpload = doc.file
        const name = doc.name
        const { filename, createReadStream } = await file
        const stream = createReadStream()
        const content = await streamToString(stream)

        documents.push({
          name,
          file: {
            id: uuidv4(),
            name: filename,
            content,
          },
        })
      }
      formSubmissions.push({
        id: uuidv4(),
        firstName,
        lastName,
        documents,
      })

      return true
    },
  },

  Upload: GraphQLUpload,
}

const app = express()
const httpServer = http.createServer(app)

const server = new ApolloServer({
  typeDefs,
  resolvers,
  plugins: [
    ApolloServerPluginDrainHttpServer({ httpServer }),
    {
      requestDidStart() {
        return Promise.resolve({
          willSendResponse(requestContext) {
            const { response } = requestContext
            // Augment response with an extension, as long as the operation
            // actually executed. (The `kind` check allows you to handle
            // incremental delivery responses specially.)
            if (
              response.body.kind === 'single' &&
              'data' in response.body.singleResult
            ) {
              response.http.headers.set('set-cookie', 'foobar=my-cookie-value')
            }
            return Promise.resolve()
          },
        })
      },
    },
    BASIC_LOGGING,
  ],
})
await server.start()

app.use(
  '/',
  cors(),
  bodyParser.json(),
  graphqlUploadExpress(),
  expressMiddleware(server, {
    context: async ({ req }) => {
      const headerClient = req.headers['x-nuxt-header-client']
      const headerServer = req.headers['x-nuxt-header-server']
      const token = req.headers.authentication || ''
      if (token !== 'server-token')
        throw new GraphQLError('you must be logged in to query this schema', {
          extensions: {
            code: 'UNAUTHENTICATED',
          },
        })
      return Promise.resolve({
        headerClient,
        headerServer,
        headers: req.headers,
      })
    },
  }),
)

await new Promise((resolve) => httpServer.listen({ port: 4000 }, () => resolve))

console.log(`ðŸš€ Server ready at http://localhost:4000/`)
