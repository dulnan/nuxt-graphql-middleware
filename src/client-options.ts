import type { GraphqlClientOptions, ContextType } from './runtime/types'

export function defineGraphqlClientOptions<T extends ContextType>(
  options: GraphqlClientOptions<T>,
): GraphqlClientOptions<T> {
  return options
}

export type { GraphqlClientOptions }
