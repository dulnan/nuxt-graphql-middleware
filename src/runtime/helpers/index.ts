/**
 * Type check for falsy values.
 *
 * Used as the callback for array.filter, e.g.
 * items.filter(falsy)
 */
export function falsy<T>(value: T): value is NonNullable<T> {
  return value !== null && value !== undefined
}
