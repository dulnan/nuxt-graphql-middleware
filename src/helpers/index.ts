import { existsSync, promises as fsp } from 'node:fs'
import { oldVisit } from '@graphql-codegen/plugin-helpers'
import { resolveFiles, resolveAlias, useLogger } from '@nuxt/kit'
import { resolve } from 'pathe'
import type { Resolver } from '@nuxt/kit'
import { inlineImportsWithLineToImports } from './fragment-import'
import { validateGraphQlDocuments } from '@graphql-tools/utils'
import { loadSchema } from '@graphql-tools/load'
import type {
  GraphQLSchema,
  GraphQLError,
  DocumentNode,
  OperationDefinitionNode,
  FragmentDefinitionNode,
} from 'graphql'
import { parse, Source, print, visit, Kind } from 'graphql'
import { falsy } from '../runtime/helpers'
import { generateSchema, generateTemplates } from './../codegen'
import { type GraphqlMiddlewareDocument } from './../types'
import { type ModuleOptions } from './../module'
import { logDocuments } from './reporter'

export const logger = useLogger('nuxt-graphql-middleware')

export const defaultOptions: ModuleOptions = {
  codegenConfig: {
    exportFragmentSpreadSubTypes: true,
    preResolveTypes: true,
    skipTypeNameForRoot: true,
    skipTypename: true,
    useTypeImports: true,
    onlyOperationTypes: true,
    namingConvention: {
      enumValues: 'change-case-all#upperCaseFirst',
    },
  },
  downloadSchema: true,
  schemaPath: '~~/schema.graphql',
  serverApiPrefix: '/api/graphql_middleware',
  graphqlEndpoint: '',
  debug: false,
  includeComposables: true,
  documents: [],
  devtools: true,
}

/**
 * Import and inline fragments in GraphQL documents.
 */
export function inlineFragments(source: string, resolver: any): string {
  return inlineImportsWithLineToImports(source, {
    resolveImport(identifier: string) {
      return resolver(identifier)
    },
    resolveOptions: {
      basedir: './',
    },
  }).inlineImports
}

/**
 * Validate the module options.
 */
export function validateOptions(options: Partial<ModuleOptions>) {
  if (!options.graphqlEndpoint) {
    throw new Error('Missing graphqlEndpoint.')
  }
}

/**
 * Get the path to the GraphQL schema.
 */
export async function getSchemaPath(
  schemaPath: string,
  options: ModuleOptions,
  resolver: Resolver['resolve'],
  writeToDisk = false,
): Promise<string> {
  const dest = resolver(schemaPath)
  if (!options.downloadSchema) {
    const fileExists = await fsp
      .access(dest)
      .then(() => true)
      .catch(() => false)
    if (!fileExists) {
      logger.error(
        '"downloadSchema" is set to false but no schema exists at ' + dest,
      )
      throw new Error('Missing GraphQL schema.')
    }
    return dest
  }
  if (!options.graphqlEndpoint) {
    throw new Error('Missing graphqlEndpoint config.')
  }
  await generateSchema(options, dest, writeToDisk)
  return dest
}

/**
 * Read the GraphQL files for the given patterns.
 */
export async function autoImportDocuments(
  patterns: string[] = [],
  srcResolver: Resolver['resolve'],
): Promise<GraphqlMiddlewareDocument[]> {
  if (!patterns.length) {
    return Promise.resolve([])
  }
  const files = (
    await resolveFiles(srcResolver(), patterns, {
      followSymbolicLinks: false,
    })
  ).filter((path) => {
    return !path.includes('schema.gql') && !path.includes('schema.graphql')
  })

  return Promise.all(
    files.map((filename) => {
      return fsp.readFile(filename).then((v) => {
        const content = v.toString().trim()
        return {
          content,
          filename,
        }
      })
    }),
  )
}

function inlineNestedFragments(
  document: string,
  fragmentMap: Record<string, string>,
): string {
  const parsed = parse(document)
  const fragmentsToInline: Set<string> = new Set()

  // Collect all fragment spreads in the document
  visit(parsed, {
    FragmentSpread(node) {
      fragmentsToInline.add(node.name.value)
    },
  })

  // Inline fragments recursively
  fragmentsToInline.forEach((fragmentName) => {
    const fragment = fragmentMap[fragmentName]
    if (fragment) {
      document += '\n' + fragment
      const nestedFragmentNames = new Set<string>()
      visit(parse(fragment), {
        FragmentSpread(node) {
          nestedFragmentNames.add(node.name.value)
        },
      })
      nestedFragmentNames.forEach((nestedFragmentName) => {
        if (!fragmentsToInline.has(nestedFragmentName)) {
          fragmentsToInline.add(nestedFragmentName)
          const nestedFragment = fragmentMap[nestedFragmentName]
          if (nestedFragment) {
            document += '\n' + nestedFragment
          }
        }
      })
    }
  })

  return document
}

export async function buildDocuments(
  providedDocuments: string[] = [],
  autoImportPatterns: string[],
  resolver: Resolver['resolve'],
  autoInlineFragments: boolean,
): Promise<GraphqlMiddlewareDocument[]> {
  const documents = await autoImportDocuments(autoImportPatterns, resolver)
    .then((importedDocuments) => [
      ...importedDocuments,
      ...providedDocuments.map((content) => ({
        content,
        filename: 'nuxt.config.ts',
      })),
    ])
    .then((documents) => {
      // If auto inlining is enabled, we can skip.
      if (autoInlineFragments) {
        return documents
      }
      return documents
        .map((v) => {
          try {
            return {
              content: inlineFragments(v.content, resolveAlias),
              filename: v.filename,
            }
          } catch (e) {
            logger.error(e)
            logger.error(
              'Failed to inline fragments for document: ' + v.filename,
            )
          }
          return null
        })
        .filter(falsy)
    })

  if (!autoInlineFragments) {
    return documents
  }

  const fragmentMap: Record<string, string> = {}
  documents.forEach((doc) => {
    const parsed = parse(doc.content)
    visit(parsed, {
      FragmentDefinition(node) {
        fragmentMap[node.name.value] = print(node)
      },
    })
  })

  documents.forEach((doc) => {
    doc.content = inlineNestedFragments(doc.content, fragmentMap)
  })

  return documents
}

export function parseDocument(
  document: GraphqlMiddlewareDocument,
  rootDir: string,
): DocumentNode {
  let name = document.filename ? document.filename.replace(rootDir, '') : ''
  if (name.charAt(0) === '/') {
    name = name.slice(1)
  }
  const source = new Source(document.content, name)
  return parse(source)
}

export function validateDocuments(
  schema: GraphQLSchema,
  documents: GraphqlMiddlewareDocument[],
  rootDir: string,
): GraphqlMiddlewareDocument[] {
  const validated: GraphqlMiddlewareDocument[] = []

  for (let i = 0; i < documents.length; i++) {
    const document: GraphqlMiddlewareDocument = { ...documents[i]! }
    if (document.filename) {
      document.relativePath = document.filename.replace(rootDir + '/', '')
    }
    try {
      const node = parseDocument(document, rootDir)
      document.content = print(node)
      document.errors = validateGraphQlDocuments(schema, [
        node,
      ]) as GraphQLError[]

      const operation = node.definitions.find(
        (v) => v.kind === 'OperationDefinition',
      ) as OperationDefinitionNode | undefined
      if (operation) {
        document.name = operation.name?.value
        document.operation = operation.operation
      } else {
        document.name = document.relativePath
      }

      document.isValid = document.errors.length === 0
    } catch (e) {
      document.errors = [e as GraphQLError]
      document.isValid = false
    }

    document.id = [document.operation, document.name, document.filename]
      .filter(Boolean)
      .join('_')

    validated.push(document)

    if (!document.isValid) {
      break
    }
  }

  return validated
}

/**
 * Parses the given document body, removes all operations except the one given
 * as the second argument and removes all fragments not used by the operation.
 */
function cleanGraphqlDocument(
  graphqlContent: string,
  operationName: string,
): DocumentNode {
  const document = parse(graphqlContent)

  let selectedOperation: OperationDefinitionNode | null = null
  const fragments: { [key: string]: FragmentDefinitionNode } = {}
  const usedFragments: Set<string> = new Set()

  visit(document, {
    OperationDefinition(node) {
      if (node.name?.value === operationName) {
        selectedOperation = node
      }
    },
    FragmentDefinition(node) {
      fragments[node.name.value] = node
    },
  })

  if (!selectedOperation) {
    throw new Error(`Operation named "${operationName}" not found`)
  }

  visit(selectedOperation, {
    FragmentSpread(node) {
      usedFragments.add(node.name.value)
    },
  })

  let hasNewFragments = true
  while (hasNewFragments) {
    hasNewFragments = false
    for (const fragmentName of usedFragments) {
      visit(fragments[fragmentName]!, {
        FragmentSpread(node) {
          if (!usedFragments.has(node.name.value)) {
            usedFragments.add(node.name.value)
            hasNewFragments = true
          }
        },
      })
    }
  }

  return {
    kind: Kind.DOCUMENT,
    definitions: [
      selectedOperation,
      ...Array.from(usedFragments).map(
        (fragmentName) => fragments[fragmentName]!,
      ),
    ],
  }
}

/**
 * Generates the TypeScript definitions and documents files.
 */
export async function generate(
  options: ModuleOptions,
  schemaPath: string,
  resolver: Resolver['resolve'],
  rootDir: string,
  logEverything = false,
) {
  const schemaContent = await fsp.readFile(schemaPath).then((v) => v.toString())
  const schema = await loadSchema(schemaContent, { loaders: [] })

  const documents = await buildDocuments(
    options.documents,
    options.autoImportPatterns as string[],
    resolver,
    !!options.autoInlineFragments,
  )

  const validated = validateDocuments(schema, documents, rootDir)

  const extracted: GraphqlMiddlewareDocument[] = validated.filter(
    (v) => !v.operation,
  )

  for (let i = 0; i < validated.length; i++) {
    const v = validated[i]
    if (!v) {
      continue
    }
    if (v.isValid) {
      try {
        const node = parse(v.content)
        oldVisit(node, {
          enter: {
            OperationDefinition: (node: OperationDefinitionNode) => {
              if (
                node.name?.value &&
                node.loc?.source &&
                (node.operation === 'query' || node.operation === 'mutation')
              ) {
                const document = { ...v }
                const cleaned = cleanGraphqlDocument(
                  node.loc.source.body,
                  node.name.value,
                )
                const errors = validateGraphQlDocuments(schema, [cleaned])
                document.errors = document.errors || []
                document.errors.push(...errors)
                document.isValid = !document.errors.length
                document.name = node.name.value
                document.operation = node.operation
                document.content = print(cleaned)
                document.id = [
                  document.operation,
                  document.name,
                  document.filename,
                ]
                  .filter(Boolean)
                  .join('_')
                extracted.push(document)
              }
            },
          },
        })
      } catch (e) {
        logger.error(e)
        extracted.push(v)
        break
      }
    } else {
      extracted.push(v)
      break
    }
  }

  const templates = await generateTemplates(
    extracted.filter((v) => v.isValid).map((v) => v.content),
    schemaPath,
    options,
  )
  const hasErrors =
    extracted.some((v) => !v.isValid) || validated.some((v) => !v.isValid)
  if (hasErrors || logEverything) {
    logDocuments(logger, extracted, logEverything)
  }

  process.stdout.write('\n')
  logger.restoreStd()

  hasErrors
    ? logger.error('GraphQL code generation failed with errors.')
    : logger.success('Finished GraphQL code generation.')

  return {
    templates: templates.sort((a, b) => {
      return a.filename.localeCompare(b.filename)
    }),
    hasErrors,
    documents: extracted.sort((a, b) => {
      const nameA = a.name || ''
      const nameB = b.name || ''
      return nameA.localeCompare(nameB)
    }),
  }
}

export const fileExists = (
  path?: string,
  extensions = ['js', 'ts', 'mjs'],
): string | null => {
  if (!path) {
    return null
  } else if (existsSync(path)) {
    return path
  }

  const extension = extensions.find((extension) =>
    existsSync(`${path}.${extension}`),
  )

  return extension ? `${path}.${extension}` : null
}

export async function outputDocuments(
  outputPath: string,
  documents: GraphqlMiddlewareDocument[],
) {
  await fsp.mkdir(outputPath, { recursive: true })
  documents.forEach((v) => {
    if (v.operation && v.name) {
      const fileName = [v.operation, v.name, 'graphql'].join('.')
      const filePath = resolve(outputPath, fileName)
      fsp.writeFile(filePath, v.content)
    }
  })
}
