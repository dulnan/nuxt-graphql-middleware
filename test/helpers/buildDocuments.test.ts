import path from 'path'
import { describe, expect, test, vi } from 'vitest'
import { createResolver, resolveAlias } from '@nuxt/kit'
import { buildDocuments } from '../../src/helpers'

// Inside vitest, the resolveAlias function does not know about the '~' alias of nuxt.
vi.mock('@nuxt/kit', async () => {
  const kit: any = await vi.importActual('@nuxt/kit')

  return {
    ...kit,
    resolveAlias: (v: string) => {
      return kit.resolveAlias(v, {
        '~': path.resolve(__dirname, './../../playground'),
      })
    },
  }
})

describe('buildDocuments', () => {
  const srcDir = path.resolve(__dirname, './../../playground')
  const resolver = createResolver(srcDir).resolve

  test('Reads and build the documents by pattern', async () => {
    expect(
      await buildDocuments([], ['./**/*.graphql'], resolver, false).then(
        (v) => {
          return v.map((doc) => {
            doc.filename = path.basename(doc.filename!)
            return doc
          })
        },
      ),
    ).toMatchSnapshot()
  })

  test('Reads and builds the documents from nuxt.config.ts', async () => {
    expect(
      await buildDocuments(
        [
          `query foobar {
        field
      }`,
        ],
        [],
        resolver,
        false,
      ),
    ).toMatchSnapshot()
  })
})
