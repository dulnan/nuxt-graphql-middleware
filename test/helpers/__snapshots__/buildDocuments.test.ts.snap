// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildDocuments > Reads and build the documents by pattern 1`] = `
[
  {
    "content": "mutation initState {
  initState
}",
    "filename": "mutation.initState.graphql",
  },
  {
    "content": "query testClientOptions($path: String!) {
  testClientOptions(path: $path) {
    language
  }
}",
    "filename": "query.testClientOptions.graphql",
  },
  {
    "content": "query getCurrentTime {
  time: getCurrentTime
}",
    "filename": "query.getCurrentTime.graphql",
  },
  {
    "content": "query fetchOptions {
  testFetchOptions {
    headerClient
    headerServer
  }
}",
    "filename": "query.fetchOptions.graphql",
  },
  {
    "content": "query fetchOptionsComposable {
  headerValueFromComposable: getRequestHeader(
    name: "x-nuxt-header-client-from-composable"
  )
  headerValueFromPlugin: getRequestHeader(name: "x-nuxt-header-client")
}",
    "filename": "query.fetchOptionsComposable.graphql",
  },
  {
    "content": "query queryOne {
  one: users {
    ...user
  }
}

query queryTwo {
  two: users {
    ...user
  }
}

fragment foobarQuery on Query {
  foobar: users {
    id
    ...foobarUser
  }
}

fragment foobarUser on User {
  firstName
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}
fragment foobarUser on User {
  firstName
}",
    "filename": "multipleQueries.graphql",
  },
  {
    "content": "mutation deleteUser($id: Int!) {
  deleteUser(id: $id)
}",
    "filename": "mutation.deleteUser.graphql",
  },
  {
    "content": "mutation triggerError {
  triggerError
}",
    "filename": "mutation.triggerError.graphql",
  },
  {
    "content": "query getError {
  getError
}",
    "filename": "query.getError.graphql",
  },
  {
    "content": "query simulateEndpointDown {
  testFetchOptions {
    headerClient
    headerServer
  }
}",
    "filename": "query.simulateEndpointDown.graphql",
  },
  {
    "content": "mutation testFormSubmit($input: FormSubmissionInput!) {
  submitForm(input: $input)
}",
    "filename": "mutation.testFormSubmit.graphql",
  },
  {
    "content": "mutation testUpload($file: Upload!) {
  uploadFile(file: $file)
}",
    "filename": "mutation.testUpload.graphql",
  },
  {
    "content": "query getSubmissions {
  getSubmissions {
    ...formSubmission
  }
}

fragment formSubmission on FormSubmission {
  id
  firstName
  lastName
  documents {
    name
    file {
      name
      id
      content
    }
  }
}
fragment formSubmission on FormSubmission {
  id
  firstName
  lastName
  documents {
    name
    file {
      name
      id
      content
    }
  }
}",
    "filename": "query.getSubmissions.graphql",
  },
  {
    "content": "query foobar {
  users {
    ...user
  }
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "test.graphql",
  },
  {
    "content": "mutation addUser($user: UserData!) {
  createUser(user: $user) {
    ...user
  }
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "mutation.addUser.graphql",
  },
  {
    "content": "query userById($id: ID!) {
  userById(id: $id) {
    ...user
  }
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "query.user.graphql",
  },
  {
    "content": "fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "user.fragment.graphql",
  },
  {
    "content": "query users {
  users {
    ...user
  }
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "userList.graphql",
  },
  {
    "content": "fragment depthOneQuery on Query {
  one: users {
    ...depthOneUser
  }
}
fragment depthOneUser on User {
  firstName
  friends {
    ...depthTwoUser
  }
}
fragment depthTwoUser on User {
  firstName
}
fragment depthTwoUser on User {
  firstName
}",
    "filename": "fragment.depthOne.graphql",
  },
  {
    "content": "fragment depthOneUser on User {
  firstName
  friends {
    ...depthTwoUser
  }
}
fragment depthTwoUser on User {
  firstName
}",
    "filename": "fragment.depthOneUser.graphql",
  },
  {
    "content": "fragment depthTwoUser on User {
  firstName
}",
    "filename": "fragment.depthTwoUser.graphql",
  },
  {
    "content": "query highDepthOne {
  ...depthOneQuery
  one: users {
    ...depthOneUser
  }
}
fragment depthOneQuery on Query {
  one: users {
    ...depthOneUser
  }
}
fragment depthOneUser on User {
  firstName
  friends {
    ...depthTwoUser
  }
}
fragment depthTwoUser on User {
  firstName
}
fragment depthTwoUser on User {
  firstName
}",
    "filename": "query.highDepthFragments.graphql",
  },
]
`;

exports[`buildDocuments > Reads and builds the documents from nuxt.config.ts 1`] = `
[
  {
    "content": "mutation initState {
  initState
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/layouts/mutation.initState.graphql",
  },
  {
    "content": "query testClientOptions($path: String!) {
  testClientOptions(path: $path) {
    language
  }
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/[lang]/query.testClientOptions.graphql",
  },
  {
    "content": "query getCurrentTime {
  time: getCurrentTime
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/caching/query.getCurrentTime.graphql",
  },
  {
    "content": "query fetchOptions {
  testFetchOptions {
    headerClient
    headerServer
  }
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/fetch-options/query.fetchOptions.graphql",
  },
  {
    "content": "query fetchOptionsComposable {
  headerValueFromComposable: getRequestHeader(
    name: "x-nuxt-header-client-from-composable"
  )
  headerValueFromPlugin: getRequestHeader(name: "x-nuxt-header-client")
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/fetch-options/query.fetchOptionsComposable.graphql",
  },
  {
    "content": "query queryOne {
  one: users {
    ...user
  }
}

query queryTwo {
  two: users {
    ...user
  }
}

fragment foobarQuery on Query {
  foobar: users {
    id
    ...foobarUser
  }
}

fragment foobarUser on User {
  firstName
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}
fragment foobarUser on User {
  firstName
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/multipleQueries.graphql",
  },
  {
    "content": "mutation deleteUser($id: Int!) {
  deleteUser(id: $id)
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/mutation.deleteUser.graphql",
  },
  {
    "content": "mutation triggerError {
  triggerError
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/mutation/mutation.triggerError.graphql",
  },
  {
    "content": "query getError {
  getError
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/query/query.getError.graphql",
  },
  {
    "content": "query simulateEndpointDown {
  testFetchOptions {
    headerClient
    headerServer
  }
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/query/query.simulateEndpointDown.graphql",
  },
  {
    "content": "mutation testFormSubmit($input: FormSubmissionInput!) {
  submitForm(input: $input)
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/test-upload/mutation.testFormSubmit.graphql",
  },
  {
    "content": "mutation testUpload($file: Upload!) {
  uploadFile(file: $file)
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/test-upload/mutation.testUpload.graphql",
  },
  {
    "content": "query getSubmissions {
  getSubmissions {
    ...formSubmission
  }
}

fragment formSubmission on FormSubmission {
  id
  firstName
  lastName
  documents {
    name
    file {
      name
      id
      content
    }
  }
}
fragment formSubmission on FormSubmission {
  id
  firstName
  lastName
  documents {
    name
    file {
      name
      id
      content
    }
  }
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/test-upload/query.getSubmissions.graphql",
  },
  {
    "content": "query foobar {
  users {
    ...user
  }
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/test.graphql",
  },
  {
    "content": "mutation addUser($user: UserData!) {
  createUser(user: $user) {
    ...user
  }
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/user/mutation.addUser.graphql",
  },
  {
    "content": "query userById($id: ID!) {
  userById(id: $id) {
    ...user
  }
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/user/query.user.graphql",
  },
  {
    "content": "fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/user/user.fragment.graphql",
  },
  {
    "content": "query users {
  users {
    ...user
  }
}
fragment user on User {
  id
  firstName
  lastName
  email
  description
  dateOfBirth
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/pages/userList.graphql",
  },
  {
    "content": "fragment depthOneQuery on Query {
  one: users {
    ...depthOneUser
  }
}
fragment depthOneUser on User {
  firstName
  friends {
    ...depthTwoUser
  }
}
fragment depthTwoUser on User {
  firstName
}
fragment depthTwoUser on User {
  firstName
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/test-queries/auto-inline/fragment.depthOne.graphql",
  },
  {
    "content": "fragment depthOneUser on User {
  firstName
  friends {
    ...depthTwoUser
  }
}
fragment depthTwoUser on User {
  firstName
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/test-queries/auto-inline/fragment.depthOneUser.graphql",
  },
  {
    "content": "fragment depthTwoUser on User {
  firstName
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/test-queries/auto-inline/fragment.depthTwoUser.graphql",
  },
  {
    "content": "query highDepthOne {
  ...depthOneQuery
  one: users {
    ...depthOneUser
  }
}
fragment depthOneQuery on Query {
  one: users {
    ...depthOneUser
  }
}
fragment depthOneUser on User {
  firstName
  friends {
    ...depthTwoUser
  }
}
fragment depthTwoUser on User {
  firstName
}
fragment depthTwoUser on User {
  firstName
}",
    "filename": "/run/media/dulnan/Storage/development/nuxt-graphql-middleware/playground/app/test-queries/auto-inline/query.highDepthFragments.graphql",
  },
  {
    "content": "query foobar {
  field
}",
    "filename": "nuxt.config.ts[0]",
  },
]
`;
